1) Сделал контроллер UserController в котором добавил

DmitryKarpushov opened the pull request:
-Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User./
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки./

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,207 @@
+public class FilmDbStorage implements FilmDb {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final MpaService mpaService;
```
В хранилищах не могут использоваться сервисы!
Наоборот, из сервисов мы вызываем хранилища и комбинируем бизнес логику.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,207 @@
+        String sqlQuery = "SELECT FILM_ID, FILM_NAME, MPA_ID, FILM_DESCRIPTION, FILM_RELEASE_DATE, FILM_DURATION, " +
+                "FILM_RATE, FILM_RATE_AND_LIKES " +
+                "FROM FILMS WHERE FILM_ID = ?";
+        return Optional.of(jdbcTemplate.queryForObject(sqlQuery, this::mapRowToFilm, id));
```
Сразу несколько моментов...
1. Есть хорошая практика делать отступ, в виде пустой строки, перед `return`.
Это выделяет точку выхода из метода визуально.
Есть только одно исключение: если метод начался сразу с return то отступ делать не нужно.
2. действие jdbcTemplate.queryForObject главное в этом методе не прячь его, оно достойно отдельной строки кода!
3. log.info("FilmDbStorage. findById."); Добавляй в лог больше полезной инфы - входящие параметры!

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,207 @@
+        String sqlQuery = "SELECT FILM_ID, FILM_NAME, MPA_ID, FILM_DESCRIPTION, FILM_RELEASE_DATE, FILM_DURATION, " +
+                "FILM_RATE, FILM_RATE_AND_LIKES " +
+                "FROM FILMS";
+        return jdbcTemplate.query(sqlQuery, this::mapRowToFilm);
```
Выборка фильмов осуществляется не правильно. Допeщена ошибка N+1
Давай поясню...
пусть у нас есть 100 фильмов... а у каждого из них есть жанры
для того чтобы собрать всё в одно и отдать при запросе нам нужно:
- достать все фильмы
- и все жанры к ним

Если делать "в лоб" то приходиться к каждому из 100-а фильмов отдельным запросом доставать еще и жанры...
получается ОДИН запрос на получение всех 100 фильмов и еще 100 запросов чтобы к каждому фильму подтянуть жанры....

Это называется проблема `N+1`
Чтобы её избежать мы должны заморочиться и в итоге приходим к всего ДВУМ запросам, вместо 101 !

Первый запрос на все фильмы, второй на все жанры для этих фильмов.

И все эти манипуляции следует выполнять в сервисе.

А вот и лог запросов, весь он не поместился на экран! ))
![image](https://user-images.githubusercontent.com/3994663/211385334-376ceeb6-c06b-4538-99a7-87cf844118b1.png)

FanaticFiz commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,66 @@
+        foreign key (GENRE_ID) references GENRE
+);
+create table USER_LIKE_FILM(
+                               FILM_ID    INTEGER               not null,
```
тут форматирование поплыло...

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.
 Насчет ошибки в тестах еще разбираюсь,но локально никаких проблем нет..

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/FilmDbStorage.java:
```diff
@@ -0,0 +1,220 @@
+package ru.yandex.practicum.filmorate.dao.impl;
+
```
Еще один момент, по поводу комментариев при коммитах.
Не нужно писать туда так много текста! Опиши коротко что было сделано именно в этот раз и всё.

А то ты просто дублируешь туда простыню одного и того же текста! Не делай так!

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Commit:
1) Изменены запросы по фильмам.

Commit:
1) Изменены запросы по фильмам.

Commit:
1) Изменены запросы по фильмам.

Commit:
1) Изменены запросы по фильмам.

Commit:
1) Изменены запросы по фильмам.
